File: frontend/src/audit.js
// SPDX-FileCopyrightText: 2026 OpenBerat contributors
// SPDX-License-Identifier: GPL-3.0-or-later

// The sessions and access screen (ADR-0026, ADR-0028). It decides nothing:
// every endpoint it calls checks ADMIN_GROUP on the handler's first line,
// independent of the decision cache, so a non-admin gets a page that says 403
// rather than a page that hides itself. Nothing here writes — revocation is
// POST /api/admin/kill/{sub}, run deliberately from a terminal (INSTALL.md §6).
//
// Plain DOM, no framework (ADR-0027). Everything that came from the API is
// written with textContent — an admin types the application name and a user
// types the path, and this page is on the host whose session cookie is valid
// for every application under .apps.<domain> (ADR-0015).

const notice = document.getElementById('notice');

function say(text) {
  notice.textContent = text;
  notice.hidden = false;
}

// --- Feature Start ---
// 403 and an outage must not read alike. The admin endpoints answer 403 to a
// user outside ADMIN_GROUP and 503 when Postgres or Redis is unreachable;
// drawing the second as the first tells an operator their access was taken
// away, and drawing the first as the second sends them to check a database
// that is fine.
// --- Feature End ---
async function api(path) {
  const response = await fetch(path, { credentials: 'same-origin' });
  if (response.status === 403) {
    throw new Error(`You are not in the group that grants the management plane, so ${path.split('?')[0]} refused you.`);
  }
  if (!response.ok) {
    let detail = '';
    try {
      const body = await response.json();
      if (body && typeof body.error === 'string') detail = ` — ${body.error}`;
    } catch (e) {
      detail = '';
    }
    throw new Error(`${path.split('?')[0]} answered ${response.status}${detail}`);
  }
  return response.json();
}

function el(tag, className, text) {
  const node = document.createElement(tag);
  if (className) node.className = className;
  if (text !== undefined && text !== null) node.textContent = text;
  return node;
}

/// Absolute and local. A relative time ("3 minutes ago") is the wrong unit for
/// a record somebody correlates against an nginx log line.
function when(iso) {
  const at = new Date(iso);
  return Number.isNaN(at.getTime()) ? iso : at.toLocaleString();
}

/// Relative, and only on the Live tab. There the question is how stale a row
/// is, not which log line it matches — "4 h ago" answers it at a glance where
/// a timestamp has to be subtracted from now by hand. The absolute time is
/// still one hover away.
function ago(iso) {
  const at = new Date(iso);
  if (Number.isNaN(at.getTime())) return iso;
  const seconds = Math.max(0, (Date.now() - at.getTime()) / 1000);
  if (seconds < 90) return 'just now';
  if (seconds < 3600) return `${Math.round(seconds / 60)} min ago`;
  if (seconds < 86400) return `${Math.round(seconds / 3600)} h ago`;
  return `${Math.round(seconds / 86400)} d ago`;
}

/// What `<input type="datetime-local">` reads and writes: local wall clock, no
/// zone. `toISOString` is UTC, so the offset has to come off first or "today"
/// starts at the wrong hour for everyone east or west of Greenwich.
function localField(at) {
  return new Date(at - at.getTimezoneOffset() * 60000).toISOString().slice(0, 16);
}

function midnight() {
  const at = new Date();
  at.setHours(0, 0, 0, 0);
  return at;
}

/// The button that carries a subject to the Explain tab. The `sub` is the one
/// value both other tables show and the one `explain` insists on — an admin
/// has a username, and typing a UUID out of a table by hand is how the wrong
/// user gets explained.
function toExplain(sub) {
  const button = el('button', 'quiet tiny', '↗');
  button.type = 'button';
  button.title = 'Explain this subject';
  button.setAttribute('aria-label', `Explain access for ${sub}`);
  button.addEventListener('click', () => {
    document.getElementById('ex-user').value = sub;
    show('explain');
    document.getElementById('ex-host').focus();
  });
  return button;
}

// --- Tabs --------------------------------------------------------------------

const VIEWS = ['live', 'history', 'explain'];

function show(name) {
  const view = VIEWS.includes(name) ? name : 'live';
  for (const other of VIEWS) {
    document.getElementById(`view-${other}`).hidden = other !== view;
    document.getElementById(`tab-${other}`).setAttribute('aria-selected', String(other === view));
  }
  if (location.hash.slice(1) !== view) location.hash = view;
  // Live is the one view that goes stale while it is open, so it reloads
  // whenever it is shown. Deliberately not a timer: a poll left running in a
  // background tab is a SCAN of the whole Redis keyspace, for nobody.
  if (view === 'live') live();
}

for (const name of VIEWS) {
  document.getElementById(`tab-${name}`).addEventListener('click', () => show(name));
}
window.addEventListener('hashchange', () => show(location.hash.slice(1)));

// --- Live --------------------------------------------------------------------

const liveRows = document.getElementById('live-rows');
const liveCount = document.getElementById('live-count');

function liveRow(entry) {
  const row = document.createElement('tr');

  const subject = el('td', 'path');
  subject.append(el('span', 'mono', entry.sub));

  // Null is not "unknown identity" — it is a session that has reached no
  // application yet, which is exactly the session this list exists to show.
  const name = el('td');
  if (entry.last_seen_as) {
    name.append(el('span', null, entry.last_seen_as));
  } else {
    name.append(el('span', 'sub', 'not seen at an application yet'));
  }

  const sessions = el('td');
  sessions.append(el('span', 'tag is-allow', String(entry.sessions)));

  // The only staleness signal there is, and it is per subject rather than per
  // session: it comes from the audit record, so it is empty both for a session
  // minted a second ago and for one nobody has touched since Tuesday. Shown as
  // it is, without a threshold and without calling any row "stale" — there is
  // no data here that would justify either.
  const activity = el('td');
  if (entry.last_activity) {
    const stamp = el('span', null, ago(entry.last_activity));
    stamp.title = when(entry.last_activity);
    activity.append(stamp);
  } else {
    activity.append(el('span', 'sub', 'no request recorded'));
  }

  const action = el('td');
  action.append(toExplain(entry.sub));

  row.append(subject, name, sessions, activity, action);
  return row;
}

function live() {
  liveCount.textContent = 'Reading…';
  api('/api/admin/sessions')
    .then((list) => {
      notice.hidden = true;
      liveRows.replaceChildren(...list.map(liveRow));
      const total = list.reduce((n, entry) => n + entry.sessions, 0);
      liveCount.textContent = list.length === 0
        ? 'No session would authenticate right now.'
        : `${total} session${total === 1 ? '' : 's'} across ${list.length} subject${list.length === 1 ? '' : 's'}`;
    })
    .catch((e) => {
      console.error(e);
      liveRows.replaceChildren();
      liveCount.textContent = '';
      say(e.message);
    });
}

document.getElementById('live-refresh').addEventListener('click', live);

// --- History -----------------------------------------------------------------

const filterForm = document.getElementById('filter-form');
const since = document.getElementById('f-since');
const until = document.getElementById('f-until');
const rows = document.getElementById('rows');
const more = document.getElementById('more');
const count = document.getElementById('count');

/// The keyset cursor: (ts, id) of the last row drawn, or null on the first
/// page. Not an OFFSET — rows arrive at the head of this ordering while an
/// admin pages through it, and the retention job deletes from the tail.
let cursor = null;
let drawn = 0;

function filters() {
  const form = new FormData(filterForm);
  const query = new URLSearchParams();
  for (const name of ['actor', 'app', 'decision', 'reason']) {
    const value = (form.get(name) || '').trim();
    if (value) query.set(name, value);
  }
  // datetime-local gives a wall-clock string with no zone; the endpoint wants
  // an instant. Interpreting it as this browser's local time is what the person
  // typing it meant.
  for (const name of ['since', 'until']) {
    const value = form.get(name);
    if (!value) continue;
    const at = new Date(value);
    if (!Number.isNaN(at.getTime())) query.set(name, at.toISOString());
  }
  return query;
}

function auditRow(event) {
  const row = document.createElement('tr');

  const at = el('td');
  at.append(el('span', null, when(event.ts)));
  // The window only means something when the row stands for more than one
  // request; drawing it always would suggest every row is a range.
  if (event.count > 1) {
    at.append(el('span', 'sub', `${when(event.first_seen)} → ${when(event.last_seen)}`));
  }

  const who = el('td');
  who.append(el('span', null, event.actor_name || event.actor_sub));
  if (event.actor_name) who.append(el('span', 'sub', event.actor_sub));
  who.append(toExplain(event.actor_sub));

  const decision = el('td');
  decision.append(el('span', `tag ${event.decision === 'allow' ? 'is-allow' : 'is-deny'}`, event.decision));
  decision.append(el('span', 'sub', event.reason));

  const requests = el('td');
  requests.append(el('span', null, String(event.count)));
  if (event.distinct_path > 1) {
    requests.append(el('span', 'sub', `${event.distinct_path} distinct paths`));
  }

  const path = el('td', 'path');
  path.append(el('span', null, event.first_path));
  const trail = [event.src_ip, event.request_id].filter(Boolean).join(' · ');
  if (trail) path.append(el('span', 'sub', trail));

  row.append(at, who, el('td', null, event.application_slug), decision, requests, path);
  return row;
}

function page() {
  const query = filters();
  if (cursor) {
    query.set('before_ts', cursor.ts);
    query.set('before_id', cursor.id);
  }
  more.disabled = true;
  api(`/api/admin/audit?${query}`)
    .then((list) => {
      rows.append(...list.map(auditRow));
      drawn += list.length;
      if (list.length > 0) {
        const last = list[list.length - 1];
        cursor = { ts: last.ts, id: last.id };
      }
      // The endpoint's own default page size. A short page is the end of the
      // record, not an error.
      more.hidden = list.length < 100;
      more.disabled = false;
      count.textContent = drawn === 0
        ? 'No rows match those filters.'
        : `${drawn} row${drawn === 1 ? '' : 's'}${more.hidden ? '' : ', more below'}`;
    })
    .catch((e) => {
      console.error(e);
      more.hidden = true;
      count.textContent = '';
      say(e.message);
    });
}

function reload() {
  notice.hidden = true;
  cursor = null;
  drawn = 0;
  rows.replaceChildren();
  page();
}

// --- Feature Start ---
// The history opens on today rather than on the whole table. `audit_event` is
// designed to grow without bound and is kept for a year by default (ADR-0022),
// so an unfiltered first page is the oldest question anybody has and the
// slowest one to answer. "All time" is one click away and says what it is.
// --- Feature End ---
filterForm.addEventListener('submit', (event) => {
  event.preventDefault();
  reload();
});
document.getElementById('filter-today').addEventListener('click', () => {
  since.value = localField(midnight());
  until.value = '';
  reload();
});
document.getElementById('filter-clear').addEventListener('click', () => {
  since.value = '';
  until.value = '';
  reload();
});
more.addEventListener('click', page);

since.value = localField(midnight());
reload();

// --- Explain -----------------------------------------------------------------

const explainForm = document.getElementById('explain-form');
const explainOut = document.getElementById('explain-out');

function verdict(answer) {
  const box = el('div', `verdict ${answer.decision === 'allow' ? 'is-allow' : 'is-deny'}`);
  box.append(el('p', 'verdict-line', `${answer.decision} — ${answer.reason}`));

  const facts = el('dl', 'facts');
  const resource = answer.resource || {};
  const pairs = [
    ['Application', resource.application],
    ['Enabled', resource.enabled === undefined ? undefined : String(resource.enabled)],
    // What the rules were actually matched against. null means the URI was
    // refused before any rule was consulted, and saying so is the whole answer
    // for half the tickets this page exists to close.
    ['Matched against', resource.normalised_path === null
      ? 'nothing — the URI was refused before any rule was consulted'
      : resource.normalised_path],
  ];
  for (const [term, value] of pairs) {
    if (value === undefined) continue;
    facts.append(el('dt', null, term), el('dd', null, value));
  }
  box.append(facts);
  return box;
}

function ruleRow(rule) {
  const row = document.createElement('tr');
  if (rule.matched) row.className = rule.expired ? 'is-expired' : 'is-matched';
  row.append(
    el('td', null, rule.effect),
    el('td', null, rule.path_pattern === '' ? '(the whole application)' : rule.path_pattern),
    el('td', null, `${rule.subject_type}: ${rule.subject_id}`),
    // A rule with a null application_id applies to every application
    // (docs/05 rule 4), and an admin reading a surprising verdict needs to see
    // that the rule they are looking at is not this application's.
    el('td', null, rule.application_id === null ? 'every application' : 'this application'),
    el('td', null, rule.expires_at ? when(rule.expires_at) : '—'),
    el('td', null, rule.matched ? (rule.expired ? 'matched, expired' : 'matched') : 'no'),
  );
  return row;
}

function rulesTable(rules) {
  const scroller = el('div', 'scroller');
  const table = document.createElement('table');
  const head = document.createElement('tr');
  for (const label of ['Effect', 'Pattern', 'Subject', 'Applies to', 'Expires', 'This request']) {
    const cell = el('th', null, label);
    cell.scope = 'col';
    head.append(cell);
  }
  const thead = document.createElement('thead');
  thead.append(head);
  const body = document.createElement('tbody');
  body.append(...rules.map(ruleRow));
  table.append(thead, body);
  scroller.append(table);
  return scroller;
}

explainForm.addEventListener('submit', (event) => {
  event.preventDefault();
  const form = new FormData(explainForm);
  // groups is sent even when empty. The endpoint requires the parameter and
  // refuses to guess: answering without it would drop every group rule and
  // report a denial that would not happen.
  const query = new URLSearchParams({
    user: form.get('user').trim(),
    groups: form.get('groups').trim(),
    host: form.get('host').trim(),
    path: form.get('path'),
  });
  explainOut.replaceChildren(el('p', 'hint', 'Asking…'));
  api(`/api/admin/explain?${query}`)
    .then((answer) => {
      const out = [verdict(answer)];
      if (answer.rules.length === 0) {
        out.push(el('p', 'prose', 'No entitlement row applies to that user on that application. A user with no rule is denied — that is a decision, not an absence.'));
      } else {
        out.push(el('p', 'prose', 'Every row the decision walked, in the order it walked them. A deny that matches wins over any allow.'));
        out.push(rulesTable(answer.rules));
      }
      explainOut.replaceChildren(...out);
    })
    .catch((e) => {
      console.error(e);
      explainOut.replaceChildren(el('p', 'refusal', e.message));
    });
});

// --- The header, shared with the portal --------------------------------------

api('/api/me')
  .then((me) => {
    document.getElementById('whoami').textContent = `Signed in as ${me.username}`;
  })
  .catch(() => {});

const signout = document.getElementById('signout');
signout.addEventListener('click', (event) => {
  event.preventDefault();
  fetch('/api/logout', { method: 'POST', credentials: 'same-origin' })
    .catch((e) => console.error(e))
    .finally(() => { location.href = signout.href; });
});

show(location.hash.slice(1));

File: backend/src/audit.rs
// SPDX-FileCopyrightText: 2026 OpenBerat contributors
// SPDX-License-Identifier: GPL-3.0-or-later

// The three admin screens that only read: the audit record, who is signed in
// (ADR-0028) and why a decision came out the way it did. None of them writes
// anything — no row, no cache entry, no Redis key — so asking a question here
// can never change the answer to it. They are routed and guarded in `admin.rs`
// like every other management-plane endpoint.

use crate::admin::{bad_request, failed};
use crate::api::Ctx;
use crate::policy;
use crate::store;
use axum::Json;
use axum::extract::{Query, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::net::IpAddr;
use std::sync::Arc;
use uuid::Uuid;

/// One `audit_event` row as the admin reads it. Every column, because the
/// summary columns are the row — a viewer showing only the decision hides that
/// it stands for 50,000 requests (docs/02, "Audit granularity").
#[derive(Serialize, sqlx::FromRow)]
struct AuditRow {
    id: Uuid,
    ts: DateTime<Utc>,
    actor_sub: String,
    actor_name: Option<String>,
    application_id: Option<Uuid>,
    application_slug: String,
    decision: String,
    reason: String,
    count: i32,
    first_seen: DateTime<Utc>,
    last_seen: DateTime<Utc>,
    distinct_path: i32,
    first_path: String,
    src_ip: Option<IpAddr>,
    request_id: Option<String>,
}

#[derive(Deserialize)]
#[serde(rename_all = "lowercase")]
enum AuditDecision {
    Allow,
    Deny,
}

// --- Feature Start ---
// Every rule below refuses the one answer an audit viewer must never give: a
// list that is quietly not the list that was asked for. `deny_unknown_fields`
// turns a mistyped filter name into a 400 rather than dropping it — a dropped
// filter widens the result, and the admin then reads "these are the denials"
// off a page with allows on it. `decision` is an enum for the same reason. And
// the page cursor is a keyset, `(ts, id)` from the last row shown, not an
// OFFSET: rows arrive at the head of this ordering while an admin pages through
// it, so OFFSET repeats page one's rows on page two, and once the retention job
// starts deleting from the tail (N-04) it skips rows instead.
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct AuditQuery {
    /// The `actor_sub` the kill switch takes or the `actor_name` a person reads
    /// off a ticket — whichever of the two the admin happens to have.
    actor: Option<String>,
    app: Option<String>,
    decision: Option<AuditDecision>,
    reason: Option<String>,
    /// Inclusive; `until` is exclusive, so consecutive windows neither overlap
    /// nor leave a gap.
    since: Option<DateTime<Utc>>,
    until: Option<DateTime<Utc>>,
    before_ts: Option<DateTime<Utc>>,
    before_id: Option<Uuid>,
    limit: Option<i64>,
}

pub(crate) async fn list_audit(
    State(ctx): State<Arc<Ctx>>,
    Query(q): Query<AuditQuery>,
) -> Response {
    if q.before_ts.is_some() != q.before_id.is_some() {
        return bad_request("before_ts and before_id are one cursor: pass both or neither");
    }
    // A cap, not a preference: this table is designed to grow without bound and
    // an unbounded LIMIT is a self-DoS an admin can type by accident.
    let limit = q.limit.unwrap_or(100).clamp(1, 1000);
    let decision = q.decision.as_ref().map(|d| match d {
        AuditDecision::Allow => "allow",
        AuditDecision::Deny => "deny",
    });
    // ponytail: matching `actor` against either column defeats
    // audit_event_actor_idx, which leads on actor_sub alone. The ts index still
    // bounds the common case — a recent page — and the alternative is two
    // parameters for one question the admin can only answer one way. Split them
    // when a search far down the table is measurably slow.
    let found: Result<Vec<AuditRow>, _> = sqlx::query_as(
        "select id, ts, actor_sub, actor_name, application_id, application_slug,
                decision, reason, count, first_seen, last_seen, distinct_path,
                first_path, src_ip, request_id
           from audit_event
          where ($1::text is null or actor_sub = $1 or actor_name = $1)
            and ($2::text is null or application_slug = $2)
            and ($3::text is null or decision = $3)
            and ($4::text is null or reason = $4)
            and ($5::timestamptz is null or ts >= $5)
            and ($6::timestamptz is null or ts < $6)
            and ($7::timestamptz is null or (ts, id) < ($7, $8::uuid))
          order by ts desc, id desc
          limit $9",
    )
    .bind(&q.actor)
    .bind(&q.app)
    .bind(decision)
    .bind(&q.reason)
    .bind(q.since)
    .bind(q.until)
    .bind(q.before_ts)
    .bind(q.before_id)
    .bind(limit)
    .fetch_all(&ctx.pool)
    .await;
    match found {
        Ok(rows) => Json(rows).into_response(),
        Err(e) => failed("list_audit", "-", e),
    }
}
// --- Feature End ---

/// One signed-in subject, as `GET /api/admin/sessions` reports it (ADR-0028).
/// `sessions` counts session keys that still exist, not set members.
#[derive(Serialize)]
struct LiveSession {
    sub: String,
    sessions: usize,
    /// From the audit record, and null for a subject it has never seen — which
    /// is the portal-only session ADR-0019 exists for. The index holds a `sub`
    /// and nothing else, and the name inside the session is behind the cookie
    /// secret this backend deliberately never holds (`session.rs`).
    last_seen_as: Option<String>,
    last_activity: Option<DateTime<Utc>>,
}

// --- Feature Start ---
// Who is signed in (ADR-0028), read out of the kill-switch index. Read-only in
// both directions: nothing is written to Redis, not even to prune a dead member,
// and no route on this screen revokes anything — revocation stays
// `POST /api/admin/kill/{sub}`, which an operator runs deliberately.
// --- Feature End ---
pub(crate) async fn list_sessions(State(ctx): State<Arc<Ctx>>) -> Response {
    let live = match ctx.index.live().await {
        Ok(live) => live,
        Err(e) => {
            tracing::error!(actor = "-", action = "list_sessions", outcome = "error", error = %e, "admin");
            return StatusCode::SERVICE_UNAVAILABLE.into_response();
        }
    };

    let subs: Vec<String> = live.iter().map(|(sub, _)| sub.clone()).collect();
    // The most recent audit row per subject. `distinct on` needs the same
    // leading column in the ordering, which is also what audit_event_actor_idx
    // leads on.
    let seen: Vec<(String, Option<String>, DateTime<Utc>)> = match sqlx::query_as(
        "select distinct on (actor_sub) actor_sub, actor_name, ts
           from audit_event where actor_sub = any($1)
          order by actor_sub, ts desc",
    )
    .bind(&subs)
    .fetch_all(&ctx.pool)
    .await
    {
        Ok(seen) => seen,
        // The list is still worth answering without names: a subject with a
        // session and no name is the one case this endpoint exists for.
        Err(e) => {
            tracing::warn!(error = %e, "sessions: the audit record could not be read for names");
            Vec::new()
        }
    };

    let mut rows: Vec<LiveSession> = live
        .into_iter()
        .map(|(sub, sessions)| {
            let seen = seen.iter().find(|(actor, _, _)| *actor == sub);
            LiveSession {
                sub,
                sessions,
                last_seen_as: seen.and_then(|(_, name, _)| name.clone()),
                last_activity: seen.map(|(_, _, ts)| *ts),
            }
        })
        .collect();
    // Most recently active first; a subject the audit record has never seen
    // sorts last rather than being dropped.
    rows.sort_by(|a, b| {
        b.last_activity
            .cmp(&a.last_activity)
            .then_with(|| a.sub.cmp(&b.sub))
    });
    Json(rows).into_response()
}

/// `GET /api/admin/explain?user&groups&host&path` — the decision the PEP would
/// reach for that request, and the rules it walked to get there.
///
/// Read-only: it fills no cache entry, writes no audit row and derives no
/// identity from the caller, so asking why a user was denied cannot itself
/// change what happens to them next. The verdict is `policy::decide`'s own
/// (`policy::explain` annotates, it does not decide) over the rows store's
/// `applicable!` hands the decision path — a screen answering differently from
/// the PEP would send an admin to fix the wrong rule. The one disagreement left
/// is deliberate: this reads the table while the PEP may still be serving a
/// cache entry, so for up to `cache::TTL` after a rule change the explanation
/// is right and the URL is stale. Reading the cache instead would make it
/// explain a decision that is about to stop being true.
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct ExplainQuery {
    /// The Keycloak `sub`, the same value the kill switch takes and the audit
    /// record's `actor_sub` column holds — not the username. It is echoed back
    /// so an admin who passed the wrong one can see that they did.
    user: String,
    /// Comma-separated, as the token carries them.
    groups: Option<String>,
    /// The `external_hostname`, which is how nginx picks the application.
    host: String,
    /// As the client would send it, query string and all: half the tickets this
    /// endpoint answers are a path that normalised into something else.
    path: String,
}

pub(crate) async fn explain(
    State(ctx): State<Arc<Ctx>>,
    Query(q): Query<ExplainQuery>,
) -> Response {
    // --- Feature Start ---
    // Required, not defaulted to none. The backend keeps no directory of its
    // own, so a missing `groups` cannot be filled in — and answering anyway
    // drops every ad_group rule and reports `no_matching_grant` for a user who
    // has access. That is the one answer this endpoint must never give.
    // --- Feature End ---
    let Some(groups) = q.groups else {
        return bad_request(
            "groups is required (comma-separated, empty for a user in none): \
             the backend holds no directory, and explaining without them drops \
             every group rule and reports a denial that would not happen",
        );
    };
    let groups: Vec<String> = groups
        .split(',')
        .map(str::trim)
        .filter(|g| !g.is_empty())
        .map(str::to_owned)
        .collect();

    // An admin pastes a URL's authority, and the schema stores neither a port
    // nor an uppercase letter.
    let host = q
        .host
        .split(':')
        .next()
        .unwrap_or_default()
        .to_ascii_lowercase();
    let found: Result<Option<(Uuid, String, bool)>, _> =
        sqlx::query_as("select id, slug, enabled from application where external_hostname = $1")
            .bind(&host)
            .fetch_optional(&ctx.pool)
            .await;
    let (application_id, slug, enabled) = match found {
        Ok(Some(row)) => row,
        Ok(None) => {
            return (
                StatusCode::NOT_FOUND,
                Json(serde_json::json!({
                    "error": "no application answers on that hostname",
                    "host": host,
                })),
            )
                .into_response();
        }
        Err(e) => return failed("explain", "-", e),
    };

    let traced = match store::traced_rules_for(&ctx.pool, application_id, &q.user, &groups).await {
        Ok(traced) => traced,
        Err(e) => return failed("explain", "-", e),
    };
    let rules: Vec<policy::Rule> = traced.iter().map(|t| t.rule.clone()).collect();
    let trace = policy::explain(enabled, &rules, &q.path, Utc::now());
    let (decision, reason) = trace.decision.as_pair();

    let walked: Vec<serde_json::Value> = traced
        .iter()
        .zip(&trace.rules)
        .map(|(row, verdict)| {
            serde_json::json!({
                "id": row.id,
                // Null is the wildcard of docs/05 rule 4: this rule applies to
                // every application, not only the one being explained.
                "application_id": row.application_id,
                "subject_type": row.subject_type,
                "subject_id": row.subject_id,
                "effect": match row.rule.effect {
                    policy::Effect::Allow => "allow",
                    policy::Effect::Deny => "deny",
                },
                "path_pattern": row.rule.path_pattern,
                "expires_at": row.rule.expires_at,
                "matched": verdict.matched,
                "expired": verdict.expired,
            })
        })
        .collect();

    Json(serde_json::json!({
        "subject": { "sub": q.user, "groups": groups },
        "resource": {
            "host": host,
            "application": slug,
            "enabled": enabled,
            "path": q.path,
            // What the rules were actually matched against. `null` means the
            // URI was refused before any rule was consulted.
            "normalised_path": trace.path,
        },
        "decision": decision,
        "reason": reason,
        "rules": walked,
    }))
    .into_response()
}
